### ISSUE TITLE
Detr models crashes when changing the num_queries parameter in the config

### ISSUE THREAD

**Issue by erickrf** [huggingface/transformers#28865]
### System Info

- `transformers` version: 4.36.2
- Platform: Linux-5.15.133+-x86_64-with-glibc2.35
- Python version: 3.10.10
- Huggingface_hub version: 0.20.2
- Safetensors version: 0.4.1
- Accelerate version: 0.26.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes, Tesla T4
- Using distributed or parallel set-up in script?: No

### Who can help?

@amyeroberts

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Load the model with a custom `num_queries` hyperparameter.
    ```
    id2label = {0: 'Test'}
    label2id = {'Test': 0}
    model_name = "facebook/detr-resnet-50"
    image_processor = AutoImageProcessor.from_pretrained(model_name)
    detr = DetrForObjectDetection.from_pretrained(
        model_name,
        id2label=id2label,
        label2id=label2id,
        ignore_mismatched_sizes=True,
        num_queries=5
    )
    ```
2. Train (or just run the forward pass with an input containing `labels`)

### Expected behavior

I would expect the model to run as normal.

I am fine tuning the model in a custom dataset which should not have more than a couple of objects per image, and expected the number of queries to have no impact other than limiting the maximum number of objects found.

**Comment by amyeroberts** [comment]
Hi @erickrf, thanks for raising this issue! 

Could you provide some more information about the crashing behaviour? Specifically, are you seeing any error messages, or is the processor just killed? 

Could you provide a minimal code snippet we can run to reproduce the error e.g. with a sample of data being passed to the model with e.g. a public dataset?

**Comment by erickrf** [comment]
Sure! I basically get the error mentioned above. 

This snippet can replicate the problem (it's rather long but from the tutorial on object detection):

```
from transformers import DetrImageProcessor, DetrForObjectDetection, TrainingArguments, Trainer
from datasets import load_dataset
import numpy as np

cppe5 = load_dataset("cppe-5")
categories = cppe5['train'].features['objects'].feature['category'].names

id2label = {index: x for index, x in enumerate(categories, start=0)}
label2id = {v: k for k, v in id2label.items()}

model_name = "facebook/detr-resnet-50"
image_processor = DetrImageProcessor.from_pretrained(model_name)
detr = DetrForObjectDetection.from_pretrained(
    model_name,
    id2label=id2label,
    label2id=label2id,
    ignore_mismatched_sizes=True,
    num_queries=5
)

def formatted_anns(image_id, category, area, bbox):
    annotations = []
    
    for i in range(0, len(category)):
        new_ann = {
            "image_id": image_id,
            "category_id": category[i],
            "isCrowd": 0,
            "area": area[i],
            "bbox": list(bbox[i]),
        }
        annotations.append(new_ann)

    return annotations


def transform_aug_ann(examples):
    image_ids = examples["image_id"]
    images, bboxes, area, categories = [], [], [], []
    
    for image, objects in zip(examples["image"], examples["objects"]):
        image = np.array(image.convert("RGB"))[:, :, ::-1]

        area.append(objects["area"])
        images.append(image)
        bboxes.append(objects["bbox"])
        categories.append(objects["category"])

    targets = [
        {"image_id": id_, "annotations": formatted_anns(id_, cat_, ar_, box_)}
        for id_, cat_, ar_, box_ in zip(image_ids, categories, area, bboxes)
    ]

    return image_processor(images=images, annotations=targets, return_tensors="pt")


def collate_fn(batch):
    pixel_values = [item["pixel_values"] for item in batch]
    encoding = image_processor.pad(pixel_values, return_tensors="pt")
    labels = [item["labels"] for item in batch]
    batch = {}
    batch["pixel_values"] = encoding["pixel_values"]
    batch["pixel_mask"] = encoding["pixel_mask"]
    batch["labels"] = labels
    return batch

cppe5["train"] = cppe5["train"].with_transform(transform_aug_ann)

training_args = TrainingArguments(
    output_dir="model/tests",
    per_device_train_batch_size=4,
    num_train_epochs=10,
    fp16=False,
    save_steps=200,
    logging_steps=200,
    learning_rate=1e-5,
    weight_decay=1e-4,
    save_total_limit=1,
    remove_unused_columns=False,
)
trainer = Trainer(
    model=detr,
    args=training_args,
    data_collator=collate_fn,
    train_dataset=cppe5["train"],
    tokenizer=image_processor,
)
trainer.train()
```

**Comment by Isalia20** [comment]
I have encountered this problem as well. When trying to change num_queries parameter it sometimes gives NAs and even worse it is unable to train. To try it out and test everything before I ran it on the whole dataset, I tried to overfit on a single image(just giving it the same image and targets on each run) but it couldn't do it in 5000 steps. Num_queries=100 worked like a charm both when starting from pretrained or without pretrained(again overfitting on a single image).

**Comment by Isalia20** [comment]
Also I found out that using a smaller learning rate fixed the Nan issue

**Comment by erickrf** [comment]
I have looked a bit more attentively into the original [DETR paper](https://arxiv.org/pdf/2005.12872.pdf), and it says (Section 3.1):

> DETR infers a fixed-size set of N predictions, in a single pass through the
decoder, where N is set to be significantly larger than the typical number of
objects in an image.

I couldn't find any analysis of the impact of this number `N`, but now I see that lowering it so much is expected to hurt the model.

Still, I would expect rather a bad performance than outright nan values.

**Comment by Isalia20** [comment]
I've looked into this quiet deeply and found that copying the weights of num_queries is useful when num_queriese is being initialized with <100 queries. So for example if it is initialized with num_queries=50, copying first 50 queries helps with training and doesn't produce nans.

@amyeroberts I can submit a PR if possible for this change(when num_queries initialized with <100 queries, to copy the first n weights). It greatly speeds up training from what I have tried.

**Comment by amyeroberts** [comment]
Hi @Isalia20, thanks for digging into the behaviour of `num_queries` and training! 

I don't think this is something we want to add in on the transformers side. The reason being that it breaks with convention of how weights are normally loaded with our models: a change in config value which causes a change in shape results in a new weight being initialized. Changing this would change assumptions about the model loading behaviour in the library. 

It sounds very useful however, please feel free to share the code or a link to an example here for the community.

**Comment by Isalia20** [comment]
I'm currently finetuning on SKU110K dataset. Once training is finished, I'll upload the model/code to HF/Github. Should I share the instructions here or is there some place else better to share?

**Comment by amyeroberts** [comment]
@Isalia20 Whereever you think is best. I'd suggest sharing here, or linking to a relevant blog / repo with example code. Another great place would be on the [forums](https://discuss.huggingface.co/).

**Comment by Isalia20** [comment]
I've released the model here:
https://huggingface.co/isalia99/detr-resnet-50-sku110k
and code is here:
https://github.com/Isalia20/DETR-finetune

**Comment by rhajou** [comment]
I am facing a similar issue but with the main difference is that the output of the model is not Nan but it does not respect the x1,y1,x2,y2 format.

Let me add this link to a similar issue found by another user on Huggingface discussions [here](https://discuss.huggingface.co/t/value-error-in-generalized-box-iou-for-conditional-detr-object-detection-training/26335/2)

Is the same solution convenient to resolve the issue?

I am trying to increase the learning rate to accelerate training. I have the following specifics:
- Number of classes = 20
- Training set/ usual Test test: we will be having a maximum of 2 objects per class

In your opinion:
- Should I change the num_queries? In this case to  less than 100 or greater than 100 ?
- Should I change to a higher learning rate than 1e-5 ?

**Comment by Isalia20** [comment]
AFAIK, the model requires x_center, y_center, width, height (in relative coordinates to image) to train

**Comment by rhajou** [comment]
@Isalia20 
But the error mentioned in this issue is mainly due to the the bboxes1 (the output of the model) and not to the bboxes2 (the target bboxes)

[not related to this issue ] 
in this case, the notebook of Niels found [here](https://colab.research.google.com/github/huggingface/notebooks/blob/main/transformers_doc/en/object_detection.ipynb#scrollTo=14ySc8VShNUy) is missing a step to convert the input from x1,y1,w,h to x_center,y_center, width_height?

**Comment by Isalia20** [comment]
Nevermind, it's actually x1, y1, w, h in relative coords and that notebook does have it correctly. My best advice would be to train with already pretrained num_queries=100 and have a small learning rate(1e-5 for the head and freeze backbone). In that case Nan issues didn't occur for me. If they still occur maybe sharing your code will help us to debug it(if possible)
